<a href="https://colab.research.google.com/github/prasann25/colab/blob/main/10_time_series_forecasting_with_tensorflow_milestone_proj3_bitpredict.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Time series forecasting fundamentals with TensorFlow + Milestone Project 3 : BitPredict

Let's write some TensorFlow code to predict the price of Bitcoin based on the historical price of Bitcoin.

**Note** : This is not a financial advice.

Links to understand some practical references:-
* All resources : https://github.com/mrdbourke/tensorflow-deep-learning

* https://dev.mrdbourke.com/tensorflow-deep-learning/

* https://eng.uber.com/forecasting-introduction/

## Get data

We're going to be using the historical price data of Bitcoin to try and predict the future price of Bitcoin, we downloaded the data from here :
https://www.coindesk.com/price/bitcoin

In [2]:
!nvidia-smi -L

GPU 0: Tesla T4 (UUID: GPU-6588e6d3-da35-2cff-9d01-cde8a623075b)


In [1]:
!wget https://raw.githubusercontent.com/mrdbourke/tensorflow-deep-learning/main/extras/BTC_USD_2013-10-01_2021-05-18-CoinDesk.csv

--2021-08-15 17:18:34--  https://raw.githubusercontent.com/mrdbourke/tensorflow-deep-learning/main/extras/BTC_USD_2013-10-01_2021-05-18-CoinDesk.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 178509 (174K) [text/plain]
Saving to: ‘BTC_USD_2013-10-01_2021-05-18-CoinDesk.csv’

BTC_USD_2013-10-01_ 100%[===================>] 174.33K  --.-KB/s    in 0.01s   

2021-08-15 17:18:35 (13.5 MB/s) - ‘BTC_USD_2013-10-01_2021-05-18-CoinDesk.csv’ saved [178509/178509]



## Importing time series with pandas

In [3]:
# Import with pandas
import pandas as pd
# Let's read in our Bitcoin data and parse the dates
df = pd.read_csv("/content/BTC_USD_2013-10-01_2021-05-18-CoinDesk.csv",
                 parse_dates=["Date"],
                 index_col=["Date"]) # parse the date column and tell pandas column 1 is a datetime
df.head()

Currency  Closing Price (USD)  ...  24h High (USD)  24h Low (USD)
Date                                      ...                               
2013-10-01      BTC            123.65499  ...       124.75166      122.56349
2013-10-02      BTC            125.45500  ...       125.75850      123.63383
2013-10-03      BTC            108.58483  ...       125.66566       83.32833
2013-10-04      BTC            118.67466  ...       118.67500      107.05816
2013-10-05      BTC            121.33866  ...       121.93633      118.00566

[5 rows x 5 columns]

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 2787 entries, 2013-10-01 to 2021-05-18
Data columns (total 5 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Currency             2787 non-null   object 
 1   Closing Price (USD)  2787 non-null   float64
 2   24h Open (USD)       2787 non-null   float64
 3   24h High (USD)       2787 non-null   float64
 4   24h Low (USD)        2787 non-null   float64
dtypes: float64(4), object(1)
memory usage: 130.6+ KB


In [5]:
# How many samples do we have ?
len(df)

2787

In [6]:
df.tail()

Currency  Closing Price (USD)  ...  24h High (USD)  24h Low (USD)
Date                                      ...                               
2021-05-14      BTC         49764.132082  ...    51448.798576   46294.720180
2021-05-15      BTC         50032.693137  ...    51578.312545   48944.346536
2021-05-16      BTC         47885.625255  ...    50690.802950   47005.102292
2021-05-17      BTC         45604.615754  ...    49670.414174   43868.638969
2021-05-18      BTC         43144.471291  ...    46622.853437   42102.346430

[5 rows x 5 columns]

We've collected the historical price of Bitcoin for the past ~8 years but there's 2787 samples.

Typically deep learning models usually like lots and lots and lots of samples (where lots and lots and lots can thousands to tens of thousands to millions)

A smaller number of samples is something you'll often run into with time series data problems.

> **Note** : The **seasonality** of a time series dataset is referred as the number of samples per year. So for our Bitcoin data, it has a seasonality of daily or value of 365 because we collect one samples per dat meaning we'll get 365 samples per year.

In [7]:
8 *365

2920